Add Index Checker annotations

kelloggm opened the pull request:
This pull request includes all of the Index Checker annotations for the JFreeChart case study. When compiling (i.e. running `mvn compile`), the Index Checker is run automatically (using the version of the checker from Maven Central).

I had to revert the old commit that added annotations in so that this can be merged cleanly. I've also removed all traces of the old `ant` build files.

The annotated version of JFreeChart is version 1.5.0.

mernst committed:
Add check-index Ant target

mernst committed:
Ignore generated files

mernst committed:
Merge github.com:jfree/jfreechart

kelloggm committed:
begin pass at annotations

kelloggm committed:
annotations

kelloggm committed:
more annotations

kelloggm committed:
more annotations

kelloggm committed:
more annotations

mernst committed:
Remove gratuitous diff against upstream

mernst committed:
Add missing import statements

kelloggm committed:
annotations on a bunch of dataset classes

kelloggm committed:
more annotations for 12/13/17

kelloggm committed:
mostly finished with data/xy package

kelloggm committed:
annotations on many packages in org/jfree/data

kelloggm committed:
improve annotations on data package. All classes now annotated; reducing remaining errors

kelloggm committed:
add suppresswarnings/switch some annotations to remove all? warnings except those caused by missing annotations in javax.swing classes

kelloggm committed:
finish annotating/suppressing warnings in data package

kelloggm committed:
begin annotating chart super-package

kelloggm committed:
additional annotations in chart packages

kelloggm committed:
fix crash

kelloggm committed:
more annotations

kelloggm committed:
remove a few more warnings

kelloggm committed:
more annotations/suppressions

kelloggm committed:
work from plane

kelloggm committed:
afternoon work

kelloggm committed:
2 warnings! The only remaining type checking warnings are super weird and need to be investigated further; everything else has either been removed or suppressed

mernst committed:
Ignore output files

kelloggm committed:
updates based on jdk changes

kelloggm committed:
Merge branch 'index' of https://github.com/kelloggm/jfreechart into index

kelloggm committed:
change comments

kelloggm committed:
update messages to clarify what is and is not a bug

kelloggm committed:
fix changes introduced b/c of jdk changes

kelloggm committed:
remove unneeded suppress warnings

kelloggm committed:
actually remove them

kelloggm committed:
minor changes

kelloggm committed:
fix bad choice

kelloggm committed:
audit establish rep invariant sws

smillst committed:
Mark warning supressions with new issue.

smillst committed:
Workaround FlowExpression parsing problem.

smillst committed:
Add LessThan annotations.

kelloggm committed:
reenable warnings suppressions in dynamic time series collection

smillst committed:
Add Stat Checkers.

smillst committed:
Merge branch 'index' of github.com:kelloggm/jfreechart into index

kelloggm committed:
readd sws to xyBarRenderer

smillst committed:
Merge branch 'index' of github.com:kelloggm/jfreechart into index

smillst committed:
Run AnnotationStatistics and JavaCodeStatistics checkers

kelloggm committed:
replace closed issue 195 with open and PR'd issue 218

kelloggm committed:
a bit of consolidation

kelloggm committed:
more consolidation

kelloggm committed:
more consolidation

kelloggm committed:
a few more consolidations in jfc

kelloggm committed:
almost finished with consolidation

kelloggm committed:
finish consolidation

panacekcz committed:
Remove suppress warnings for typetools/kelloggm#218

smillst committed:
Merge pull request #1 from panacekcz/unsuppress-218

Remove SuppressWarnings for modulo

smillst committed:
Remove Unneeded annotations.

smillst committed:
Merge branch 'index' of github.com:kelloggm/jfreechart into index

kelloggm committed:
remove some unnecessary sws

kelloggm committed:
Merge branch 'index' of https://github.com/kelloggm/jfreechart into index

kelloggm committed:
a few more annotations

kelloggm committed:
apparently an annotation is required on protected fields

kelloggm committed:
fix merge conflicts

kelloggm committed:
readd lib files that CF needs apparently

kelloggm committed:
minor fix

jfree committed:
Fix test that fails in different timezones.

jfree committed:
Fix for bug #75 - add same outline options as XYLineAndShapeRenderer.

jfree committed:
Update Javadocs for overlays.

jfree committed:
Add generics for subtitles list.

jfree committed:
Don't use buffering in ChartPanel.

kelloggm committed:
fix build

kelloggm committed:
remove old lib file, delete them from .gitignore

panacekcz commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -423,8 +433,10 @@ public static int stringToMonthCode(String s) {
             }
         }
+        @SuppressWarnings({"index", "value"}) // the loop above only ensures that 1 <= result <= 12 if s is an actual month name
+        @IntRange(from = 1, to =12) int toReturn = result;
```
Should be `@IntRange(from = -1, to = 12)`?

kelloggm replied:
No. The if loop above ensures that the value is 1 to 12 if the variable s is a valid month name, which is a precondition of the function.

panacekcz commented on src/main/java/org/jfree/chart/date/SpreadsheetDate.java:
```diff
@@ -76,7 +82,7 @@
      * @param year  the year (in the range 1900 to 9999).
      */
-    public SpreadsheetDate(int day, int month, int year) {
+    public SpreadsheetDate(@IntRange(from = 1, to = 31) int day, /*@IntRange(from = 1, to = 12)*/ int month, @IntRange(from = 1, to = 9999) int year) {
```
Annotation in comment.

panacekcz commented on src/main/java/org/jfree/chart/date/SpreadsheetDate.java:
```diff
@@ -425,16 +445,22 @@ else if (include == SerialDate.INCLUDE_SECOND) {
      * @return the serial number from the day, month and year.
      */
-    private int calcSerial(int d, int m, int y) {
+    private @IntRange(from = 2, to = 2958465) int calcSerial(/*@IntRange(from = 1, to = 31)*/ int d, /*@IntRange(from=1,to=12)*/ int m, @IntRange(from = 1900, to = 9999) int y) {
```
Annotations in comments.

panacekcz commented on src/main/java/org/jfree/data/DataUtils.java:
```diff
@@ -238,10 +241,11 @@ public static double calculateRowTotal(Values2D data, int row,
      */
-    public static Number[][] createNumberArray2D(double[][] data) {
+    @SuppressWarnings("value") // the result array is manifestly the same size as the parameter
+    public static Number @SameLen("#1") /*@PolyValue*/ [][] createNumberArray2D(double @PolyValue [][] data) {
```
Annotation in comments.

panacekcz commented on src/main/java/org/jfree/data/statistics/Regression.java:
```diff
@@ -63,7 +70,7 @@
      * @return The parameters.
      */
-    public static double[] getOLSRegression(double[][] data) {
+    public static double @ArrayLen(2) [] getOLSRegression(double /*@MinLen(2)*/ [] @MinLen(2) [] data) {
```
Annotation in comment.

panacekcz commented on src/main/java/org/jfree/data/statistics/Regression.java:
```diff
@@ -150,7 +157,7 @@
      * @return The parameters.
      */
-    public static double[] getPowerRegression(double[][] data) {
+    public static double @ArrayLen(2) [] getPowerRegression(double /*@MinLen(2)*/ [] @MinLen(2) [] data) {
```
Annotation in comments.

panacekcz commented on src/main/java/org/jfree/data/statistics/Statistics.java:
```diff
@@ -327,7 +332,7 @@ public static double getStdDev(Number[] data) {
      * @return A double array with the intercept in [0] and the slope in [1].
      */
-    public static double[] getLinearFit(Number[] xData, Number[] yData) {
+    public static double @ArrayLen(2) [] getLinearFit(Number /*@SameLen("#2")*/ [] xData, Number @SameLen("#1") [] yData) {
```
Annotation in comment.

panacekcz commented on src/main/java/org/jfree/data/statistics/Statistics.java:
```diff
@@ -439,8 +444,8 @@ public static double getCorrelation(Number[] data1, Number[] data2) {
      */
-    public static double[][] getMovingAverage(Number[] xData, Number[] yData,
-            int period) {
+    public static double[] @ArrayLen(2) [] getMovingAverage(Number /*@SameLen("#2")*/ [] xData, Number @SameLen("#1") [] yData,
```
Annotation in comment.

panacekcz commented on src/main/java/org/jfree/data/xy/DefaultHighLowDataset.java:
```diff
@@ -98,9 +100,11 @@
-            double[] high, double[] low, double[] open, double[] close,
-            double[] volume) {
+    @SuppressWarnings("samelen") // While initializing object, SameLen invariants between the fields will be broken (because one field must be initialized before the others). The annotations on the parameters guarantee the invariant holds after the constructor finishes executing.
+    public DefaultHighLowDataset(Comparable seriesKey, Date @SameLen({"#2", "#3", "#4", "#4", "#5", "#6", "#7"}) [] date,
```
Duplicate `"#4"`

panacekcz commented on src/test/java/org/jfree/chart/GanttChartTest.java:
```diff
@@ -271,7 +271,7 @@ public static IntervalCategoryDataset createDataset() {
      * @return a date.
      */
-    private static Date date(int day, int month, int year) {
+    private static Date date(@IntRange(from = 1, to = 31) int day, /*@IntRange(from = 1, to = 12)*/ int month, @IntRange(from = 1900, to = 9999) int year) {
```
Annotation in comment.

panacekcz reviewed (commented):
I just quickly looked through the diff, looks good to me, just nitpicks: found some annotations still in comments (see review comments), sometimes inconsistent whitespace (I didn't mark that).

kelloggm committed:
minor cosmetic fixes Vlastimil noticed

kelloggm committed:
add travis integration

kelloggm committed:
change timeout

kelloggm committed:
some whitespace fixes

kelloggm committed:
more whitespace fixes

kelloggm committed:
minor changes to travis ci config

kelloggm committed:
more whitespace fixes

mernst commented on .gitignore:
```diff
@@ -2,10 +2,12 @@
 /target/
 /javadoc/
 
+build/
```
Why is this change needed?

mernst commented on .gitignore:
```diff
@@ -2,10 +2,12 @@
 # Intellij Idea
-.idea
\ No newline at end of file
+.idea
```
Please don't make unnecessary whitespace changes.  It pollutes the diffs.

mernst commented on .travis.yml:
```diff
@@ -0,0 +1,5 @@
+sudo: false
+script:
+  - mvn clean
+  - travis_wait mvn compile
```
Add newline at end of files, when creating a new file.

mernst commented on .travis.yml:
```diff
@@ -0,0 +1,5 @@
+language: java
+sudo: false
+script:
+  - mvn clean
```
Why is this needed?  A Travis job always starts out from a clean clone.

mernst commented on pom.xml:
```diff
@@ -59,31 +59,90 @@
+        <dependency>
+            <groupId>org.checkerframework</groupId>
+            <artifactId>checker-qual</artifactId>
+            <version>2.5.0</version>
```
Update to latest version.

mernst commented on pom.xml:
```diff
@@ -59,31 +59,90 @@
+        <dependency>
+            <groupId>org.checkerframework</groupId>
+            <artifactId>jdk8</artifactId>
+            <version>2.5.0</version>
```
Update to latest version.

mernst commented on pom.xml:
```diff
@@ -59,31 +59,90 @@
           <id>ossrh</id>
           <url>https://oss.sonatype.org/content/repositories/snapshots</url>
         </snapshotRepository>
-    </distributionManagement>    
```
There was (gratiutous) whitespace at the end of this line, which you removed.  Please reinstate it, and review the diffs to ensure that no other whitespace changes remain.

mernst commented on pom.xml:
```diff
@@ -59,31 +59,90 @@
-                <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
-                <version>3.7.0</version>
+                <version>3.6.1</version>
```
Why did you make this change?

mernst commented on pom.xml:
```diff
@@ -59,31 +59,90 @@
-                    <source>${project.source.level}</source>
-                    <target>${project.target.level}</target>
-                    <encoding>${project.build.sourceEncoding}</encoding>                
+                    <source>1.8</source>
```
Rather than hard-coding here, why not change the version number at line 159?  That will reduce the size of the diffs and make it easier to merge in future changes.

mernst commented on src/main/java/org/jfree/chart/annotations/XYPointerAnnotation.java:
```diff
@@ -376,7 +378,7 @@ public void setArrowPaint(Paint paint) {
      */
     @Override
     public void draw(Graphics2D g2, XYPlot plot, Rectangle2D dataArea,
-            ValueAxis domainAxis, ValueAxis rangeAxis, int rendererIndex, 
```
Undo the whitespace change (you deleted the space character at the end of the line).

mernst commented on src/main/java/org/jfree/chart/annotations/XYPolygonAnnotation.java:
```diff
@@ -220,16 +222,21 @@ public void draw(Graphics2D g2, XYPlot plot, Rectangle2D dataArea,
 
         GeneralPath area = new GeneralPath();
+
+        @SuppressWarnings("index") // Length is checked above, and all functions called in-between are pure
```
Say length of what.  Also, I think it would be better to annotate the functions as pure rather than writing this `@SuppressWarnings`.  The `@Pure` annotations will be helpful to other people, and you will have fewer `@SuppressWarnings` annotations cluttering the code.

This same comment applies to other `@SuppressWarnings` about purity (there's one 3 lines below, and might be others in the codebase).

mernst commented on src/main/java/org/jfree/chart/annotations/XYPolygonAnnotation.java:
```diff
@@ -220,16 +222,21 @@ public void draw(Graphics2D g2, XYPlot plot, Rectangle2D dataArea,
                 x = domainAxis.valueToJava2D(this.polygon[i], dataArea,
                         domainEdge);
-                y = rangeAxis.valueToJava2D(this.polygon[i + 1], dataArea,
+                @SuppressWarnings("index") // polygon always has an even number of elements, and i is an odd index
```
`i` is an even index.  Same comment applies on line 250 below.

mernst commented on src/main/java/org/jfree/chart/axis/CyclicNumberAxis.java:
```diff
@@ -476,7 +479,9 @@ protected List refreshTicksHorizontal(Graphics2D g2, Rectangle2D dataArea,
                     anchor = isInverted()
                         ? TextAnchor.TOP_RIGHT : TextAnchor.BOTTOM_RIGHT;
-                    result.remove(result.size() - 1);
+                    @SuppressWarnings("index") // If cycled, there must be at least one tick in the results list already
```
This appears a number of times.  You could abstract it out as a new private method `removeLast`.
I think that would be clearer, it would reduce the size of the diffs, and it would reduce the number of warning suppressions.

mernst commented on src/main/java/org/jfree/chart/axis/DateAxis.java:
```diff
@@ -1005,7 +1007,9 @@ else if (this.tickMarkPosition == DateTickMarkPosition.MIDDLE) {
         }
         else if (DateTickUnitType.MONTH.equals(unit.getUnitType())) {
-            value = count * ((current + 1) / count) - 1;
+            @SuppressWarnings("index") // if this is a month, then current >= 1, so the below is safe
```
The diffs would be shorter and clearer if you just add one line:
```assert current >= 1 : "@AssumeAssertion(index)"; // if this is a month, then current >= 1```

mernst commented on src/main/java/org/jfree/chart/axis/MonthDateFormat.java:
```diff
@@ -163,18 +166,20 @@ public MonthDateFormat(TimeZone zone, Locale locale, int chars,
         String[] monthsFromLocale = dfs.getMonths();
         this.months = new String[12];
         for (int i = 0; i < 12; i++) {
+            @SuppressWarnings({"index", "value"}) // https://github.com/typetools/checker-framework/issues/1669
```
Write assertions on the value of `i` rather than introducing new variables.

mernst commented on src/main/java/org/jfree/chart/axis/TickUnits.java:
```diff
@@ -126,6 +128,7 @@ public TickUnit get(int pos) {
      * @return A tick unit that is larger than the supplied unit.
      */
     @Override
+    @SuppressWarnings("index") // binary search on list
```
Shouldn't binary search indices (`@SearchIndexFor`) handle this code?  If not, please make the warning suppression more specific.  The same comment applies to line 154 and maybe elsewhere.

mernst commented on src/main/java/org/jfree/chart/block/ColumnArrangement.java:
```diff
@@ -308,6 +310,7 @@ protected Size2D arrangeRF(BlockContainer container, Graphics2D g2,
      *
      * @return The size after the arrangement.
      */
+    @SuppressWarnings("index") // https://github.com/kelloggm/checker-framework/issues/152
```
I would add "@LengthIs" to the comment, before the URL, to make it easier for readers to know what the warning suppression is about.  The same comment applies to other comments that consist of just a URL:  writing one or two words to explain them is helpful.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -361,21 +367,23 @@ public static String monthCodeToString(int month) {
 
         if (shortened) {
-            months = DATE_FORMAT_SYMBOLS.getShortMonths();
+            final String[] monthsTmp = DATE_FORMAT_SYMBOLS.getShortMonths();
```
Why is this change necessary?

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -361,21 +367,23 @@ public static String monthCodeToString(int month) {
         }
         else {
-            months = DATE_FORMAT_SYMBOLS.getMonths();
+            final String[] monthsTmp = DATE_FORMAT_SYMBOLS.getMonths();
```
Why is this change necessary?

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -394,10 +402,11 @@ public static String monthCodeToString(int month, boolean shortened) {
 
         final String[] shortMonthNames = DATE_FORMAT_SYMBOLS.getShortMonths();
-        final String[] monthNames = DATE_FORMAT_SYMBOLS.getMonths();
+        @SuppressWarnings("index") // Need annotations on Date Format Symbols
```
Is this necessary?  The calls to `getMonths` above didn't need `@SuppressWarnings`.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -346,7 +352,7 @@ public static int monthCodeToQuarter(int code) {
      * @return a string representing the supplied month.
      */
-    public static String monthCodeToString(int month) {
+    public static String monthCodeToString(@IntRange(from=1, to=12) int month) {
```
Inconsistent spacing; compare to line 324 above.  Either always, or never, put spaces around `=`.  Probably always, but I don't care which you choose so long as it is consistent.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -423,8 +432,10 @@ public static int stringToMonthCode(String s) {
             }
         }
+        @SuppressWarnings({"index", "value"}) // the loop above only ensures that 1 <= result <= 12 if s is an actual month name
+        @IntRange(from = 1, to =12) int toReturn = result;
```
Here and elsewhere, write an assertion instead of introducing a new temporary variable.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -479,9 +490,10 @@ else if ((yyyy % 100) == 0) {
      * @param yyyy  the year (in the range 1900 to 9999).
      *
-     * @return the number of leap years from 1900 to the specified year.
+     * @return the number of leap years from 1900 to the specified year (if year is 9999, 1964 leap years).
```
Comment is too telegraphic.  Add a few words to clarify.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -479,9 +490,10 @@ else if ((yyyy % 100) == 0) {
      */
-    public static int leapYearCount(int yyyy) {
+    @SuppressWarnings({"index", "value"}) // imprecision wrt ranges
+    public static @IntRange(from = 0, to = 1964) int leapYearCount(@IntRange(from = 1900, to = 9999)int yyyy) {
```
Needs whitespace between ")" and "int".

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -538,14 +552,17 @@ public static SerialDate addDays(int days, SerialDate base) {
      *
      * @return a new date.
      */
+    @SuppressWarnings({"index", "value"}) // True positive, fixed upstream bug
```
Reference an issue tracker bug in upstream JFreeChart repository.
The same comment applies to other comments of this form.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -538,14 +552,17 @@ public static SerialDate addDays(int days, SerialDate base) {
         int mm = (12 * base.getYYYY() + base.getMonth() + months - 1) % 12 + 1;
-        int dd = Math.min(base.getDayOfMonth(), 
-                SerialDate.lastDayOfMonth(mm, yy));
+        int lastDayOfMonth = SerialDate.lastDayOfMonth(mm, yy);
```
What is the point of abstracting out this variable?
(Maybe there used to be a `@SuppressWarnings` on it?)
Please review the diffs and undo any unnecessary changes like this, to minimize the diffs.

mernst commented on src/main/java/org/jfree/chart/date/SerialDate.java:
```diff
@@ -581,7 +599,7 @@ public static SerialDate addYears(int years, SerialDate base) {
      * @return the latest date that falls on the specified day-of-the-week and 
      *         is BEFORE the base date.
      */
-    public static SerialDate getPreviousDayOfWeek(int targetWeekday, 
```
Don't change whitespace (there was a gratuitous trailing space here).

mernst reviewed (changes requested):
The standard Maven target is `mvn package`.
That works for upstream JFreeChart, but it fails for this pull request.
Please make `mvn package` work, or document some other command that builds all the artifacts.

I have reviewed about 20% of the changes, but many of my comments apply throughout the codebase.  Please make those changes and I will re-review.

kelloggm committed:
address a number of Mike's comments

kelloggm committed:
fix merge

kelloggm committed:
address more of Mike's commments

mernst commented on src/main/java/org/jfree/chart/ChartFrame.java:
```diff
@@ -2,7 +2,7 @@
  * ===========================================================
  *
- * (C) Copyright 2000-2016, by Object Refinery Limited and Contributors.
+ * (C) Copyright 2000-2018, by Object Refinery Limited and Contributors.
```
I don't think the copyright dates should be changed.

mernst replied:
If this is because you want to revert the JFreeChart source code to an earlier version, that should be done in a separate pull request, to make this one easier to review.

kelloggm replied:
This is different because typetools hasn't merged from upstream, but this branch has (as of about 30 minutes ago when I realized I needed to because the tests were failing!). I certainly didn't change the copyright date. I'll merge upstream into typetools/master to fix the diff.

kelloggm committed:
more changes to address Mike's comments

kelloggm committed:
add a second extra argument because Maven is bad

kelloggm committed:
minor fixes

smillst committed:
Work on Guaranteed index false positives.

1. Remove two warning suppressions that aren't required.  (One required more annotations.)
2. Added "rewrite" to 14 of the false positives that could be fixed by using Map.Entry.

smillst committed:
Remove warning suppressions that aren't needed.

smillst committed:
Convert 4 false positives into one.

Plus remove one more.

smillst committed:
Improve comment.

smillst committed:
Add type argument to iterator and @NonNeg

kelloggm committed:
Merge pull request #2 from smillst/index

Work on Guaranteed index false positives.
